### ISSUE TITLE
[Custom UI] Allow main menu music to be configured 

### ISSUE THREAD

**Issue by ForestDragon-wesnoth** [wesnoth/wesnoth#9436]
### Describe the desired feature

With 1.19.4 introducing support for custom themes (https://wiki.wesnoth.org/GUIToolkit#Custom_themes), and title screen being editable there too, I believe that the title screen theme should be configurable in the UI settings instead of being limited to cores.

So return_to_wesnoth.ogg stays as a core option, but custom UIs can set their own music, which, if specified, has higher priority than core settings.

This would allow use cases like playing main_menu.ogg for nostalgia, or the opposite - more unusual main menus that also sound different (like a custom sci-fi theme for UMC, or a more dark-fantasy-like title screen with less upbeat music), all without needing the cumbersome workaround of a whole core to just change the main theme.

EDIT: updated the post to be more accurate based on information provided by soliton

**Comment by ForestDragon-wesnoth** [comment]
cores are a much more large-scale thing than custom UI, and having a core enabled disables all add-ons that are not designed for the core.

I believe allowing GUI music settings to overwrite core settings would make a lot of sense

**Comment by soliton-** [comment]
If your core simply changes the title screen music I see no reason why it would disable all addons.

**Comment by ForestDragon-wesnoth** [comment]
Oh, seems I misunderstood that part then.

Still, both the UI and music are thematic things, like cores usually focus more on code aspects and can optionally have their own custom UI, so I still believe letting UI settings have higher priority for music than core settings is a good idea.

**Comment by ForestDragon-wesnoth** [comment]
And there is already some precedent of sound in UI settings, like sound_button_click, etc.

**Comment by soliton-** [comment]
It might make sense to let every window/dialog define its own background music.

**Comment by ForestDragon-wesnoth** [comment]
> It might make sense to let every window/dialog define its own background music.

that could potentially work, though in the short term at least the main menu should have that

**Comment by CelticMinstrel** [comment]
> If your core simply changes the title screen music I see no reason why it would disable all addons.

That makes sense, but I'm not sure if things actually work that way.

> The title screen music is specified here:

That's not the entirety of it. There's an entire playlist defined somewhere.

---

Ultimately, I think it _makes sense_ for titlescreen music to be in the core, rather than the theme. If you make a custom core, there's a good chance you don't want the regular Wesnoth music. I can definitely see an argument for it being in the theme too, mind you, as it's also an obvious thing a theme might want to customize. This seems like a tough decision…

**Comment by ForestDragon-wesnoth** [comment]
> Ultimately, I think it _makes sense_ for titlescreen music to be in the core, rather than the theme. If you make a custom core, there's a good chance you don't want the regular Wesnoth music. I can definitely see an argument for it being in the theme too, mind you, as it's also an obvious thing a theme might want to customize. This seems like a tough decision…

I think what I am proposing above should satisfy both needs well enough - if the theme does not define its own music, load music from the core, the default themes do not specify music but custom themes can have their own music types. custom themes' music takes higher priority over core settings

**Comment by CelticMinstrel** [comment]
> I think what I am proposing above should satisfy both needs well enough

Well, no, not really. It breaks down as soon as someone actually _uses_ a custom core and a custom theme together. Say someone made a Star Wars core (just as a random example; obviously it wouldn't be allowed on the add-ons server, but we can't stop people from actually _making_ it), and sets the Star Wars Main Theme as the title music. But then someone using that core decides to activate a theme that sets Cry from Elensefar as the title music. That sounds very wrong to me.

**Comment by ForestDragon-wesnoth** [comment]
@babaissarkar as the current UI dev, what do you think on this?

**Comment by babaissarkar** [comment]
> @babaissarkar as the current UI dev, what do you think on this?

I don't think it's a bad idea, though we need to decide whether the core or the theme should have higher precedence. I'm willing to implement it if we reach a consensus. (I don't have problems with either order.)

**Comment by ForestDragon-wesnoth** [comment]
if the returning_to_wesnoth.ogg music stays defined in core AND cores take priority, it would only be possible to overwrite it with other cores but not themes

So it's either a) default main music stays is core, but themes have higher priority b) default core's music is moved to default themes, cores are higher priority than themes

(personally I am in favor of option a)

**Comment by CelticMinstrel** [comment]
Also, don't forget there's also the titlescreen playlist. Where does that belong? In the core? Or in the theme? I'm not even sure which one it would be considered to be in right now.

**Comment by ForestDragon-wesnoth** [comment]
> Also, don't forget there's also the titlescreen playlist. Where does that belong? In the core? Or in the theme? I'm not even sure which one it would be considered to be in right now.

I think it would make sense that titlescreen playlists should be defined in the same places as the main title music, and follow the same priorities (otherwise, as you pointed out, we could get sci-fi music then switching to Cry from Elensefar when the sci-fi track ends)

**Comment by CelticMinstrel** [comment]
I don't understand how that solves anything…?

**Comment by babaissarkar** [comment]
> I don't understand how that solves anything…?

A third file where overlapping settings are stored. This removes the overlap between cores and themes. Though TBH I don't really know much about cores apart from what I can see in https://wiki.wesnoth.org/CoreWML. I think some or all of the settings in game_config should be exposed (such as setting background or music) in a simple manner, so that people do not have to create a core just to customize trivial stuff.
Do cores have a fallback mechanism, such that if a file does not exist in a core, it gets loaded from the default core?
In ForestDragon's case, he'll need both a core and a theme in order to fully recreate the 1.18 experience. If it could be done from the theme then that make it much simpler.
Similarly, if someone wishes to change the background, they'll also need a core. (it's possible with themes, but require some complications).

**Comment by ForestDragon-wesnoth** [comment]
> Similarly, if someone wishes to change the background, they'll also need a core. (it's possible with themes, but require some complications).

true, adding a more convenient way to edit backgrounds without fully overwriting title_screen.cfg could be nice

maybe to make it fit both the needs of Default theme and Modern theme, it would accept a comma-seperated list of backgrounds that are randomly chosen, and in Default theme's case the list is just one image?

**Comment by ForestDragon-wesnoth** [comment]
> A third file where overlapping settings are stored. This removes the overlap between cores and themes. 

I have to agree with Celmin that I am confused about the idea

**Comment by CelticMinstrel** [comment]
The things that can be customized in cores are _everything_ in the game config (as specified by the main schema file); the information at CoreWML only explains how to tell the engine that your core exists and not how to actually create a core. (It does have a few tips on the latter but no detailed information.)

**Comment by soliton-** [comment]
> > The title screen music is specified here:
> 
> That's not the entirety of it. There's an entire playlist defined somewhere.

Indeed there is 02bef46e0bf990e8e51296846a3175ecbfc3e737. That does not seem like a sensible design though. Could have instead made `title_music` in game_config a comma separated list or something. Having two places that specify music mixed up like that is confusing. IMO this should be cleaned up (regardless of how this feature request is implemented).

**Comment by cooljeanius** [comment]
Related to #2797 (possibly a dup of it?)

**Comment by ForestDragon-wesnoth** [comment]
> Related to #2797 (possibly a dup of it?)

Not a duplicate - #2797 seems more about players manually choosing tracks via UI, while #9436 is about supporting different title screen music in theme code
